The OCAMLRUN build variable

shindere opened the pull request:
This PR makes it possible to use the OCAMLRUN variable to specify
which runtime to use while building the compilers.

Mostly, it renames the CAMLRUN variable to OCAMLRUN (@dra27) and
attempts to make sure this is used consistently through all the
build system.

There may still be places to tune, especially in connection with FlexDLL,
that's one example of situation where @dra27's sharp eyes would be very
helpful.

3 remarks are due.

 1. The PR starts with something unrelated, namely the replacement of our oldish `.SUFFIXES` rules by pattern-rules in a few places. I checked that this changes absolutely nothing to the build, but this part does still have a user-visible consequence, since such rules were also used in the manual, in the part on ocamldep. I could isolate this in a dedicated PR, but frankly it didn't feel worth it (but I can still do so if asked to).
 2. In two places, I changed the runtime used: it was `runtime/ocamlrun` which was hard-coded and, after this PR, it will be `boot/ocamlrun` which will be used by default, see the relevant commit message for the details. I think this change should be painless and even consider that the former situation was kind of an error, an oversight.
 3. This PR introduces the `NEW_OCAMLRUN` variable, that defaults to `runtime/ocamlrun` because it didn't seem possible to do without that, especially during the bootstrap.

So far, I have checked that:

 * The bootstrap still works
 * `make OCAMLRUN=/foo/bar/ocamlrun` indeed uses that runtime everywhere.

shindere committed:
Build system: replace suffix rules by pattern rules

shindere committed:
Build system: rename the CAMLRUN variable to OCAMLRUN

Since ocamldoc/Makefile.best_ocamldoc was already using OCAMLRUN (rather
than CAMLRUN), it was possible to remove its definition in that file,
because the value is now initialised globally in Makefile.common.

shindere committed:
Build system: use boot/ocamlrun rather than runtime/ocamlrun in some places

In two places, the use of runtime/ocamlrun was hardcoded:

- In the root Makefile to run the top-level, i.e. make runtop

- In tools/Makefile, to build the camltex tool

This commit replaces these two hard-coded paths by a reference to the
OCAMLRUN variable, which means it changes the runtime effectively used
by default in these two places but this shouldn't make a big difference
and it means that when OCAMLRUN gets overloaded, its new value will also
be taken into account in these places, which seems to make sense and be desirable.

shindere committed:
Introduce the NEW_OCAMLRUN build variable

This variable is initialised by default to runtime/ocamlrun

Most of the time this runtime is identical to boot/ocamlrun, but not always.
During the bootstrap process, for instance, the two runtimes may differ.

With this variable it becomes possible, when calling the build system,
to override the default value, which was not possible before.

shindere committed:
Add a Changes entry

dra27 commented:
Just a quick note the failure on the full-flambda job is unrelated (and is being worked on)

shindere commented:
David Allsopp (2021/04/22 03:42 -0700):
> Just a quick note the failure on the full-flambda job is unrelated
> (and is being worked on)

Thanks! Do you need help there? Something that I can help with?

dra27 commented on Makefile:
```diff
@@ -82,7 +82,7 @@ LIBFILES=stdlib.cma std_exit.cmo *.cmi camlheader
 
 TOPINCLUDES=$(addprefix -I otherlibs/,$(filter-out %threads,$(OTHERLIBRARIES)))
-RUNTOP=./runtime/ocamlrun$(EXE) ./ocaml$(EXE) \
+RUNTOP = $(OCAMLRUN) ./ocaml$(EXE) \
```
I initially thought this was incorrect: if you've just _added_ a primitive and updated the stdlib, then you'd expect to be able to use that primitive. However, that's not correct: while you can add a new primitive to the runtime and use it without _bootstrapping_, you can't do that without repeating _coldstart_.

Therefore, `ocaml$(EXE)` is produced using the `BOOT_OCAMLC` and should therefore be running on `$(OCAMLRUN)` and not `runtime/ocamlrun$(EXE)`.

Aha, but if you're working on a fix in the runtime with no changes to bytecode and primitives, might you expect `make runtop` to use the runtime you're working on, rather than one you booted with? Well, at the moment that doesn't matter because `make runtop` begins with a `coldstart`. However, cf. https://github.com/ocaml/ocaml/pull/10114#pullrequestreview-560715273.

I think the conclusion of my ramble is that above the `$(MAKE) coldstart)` in the `runtop` recipe it might be worth adding a comment that `$(RUNTOP)` could want changing to use `runtime/ocamlrun` if we ever _stop_ doing `coldstart` at the start of `runtop`.

dra27 commented on tools/Makefile:
```diff
@@ -319,7 +319,7 @@ caml_tex := caml-tex$(EXE)
 $(caml_tex): INCLUDES += $(addprefix -I $(ROOTDIR)/otherlibs/,str $(UNIXLIB))
 $(caml_tex): $(caml_tex_files)
-	$(ROOTDIR)/runtime/ocamlrun$(EXE) $(ROOTDIR)/ocamlc$(EXE) -nostdlib \
+	$(OCAMLRUN) $(ROOTDIR)/ocamlc$(EXE) -nostdlib \
```
Likewise, this is correct but possibly worth a comment clarifying (especially if what I write is wrong!). `$(ROOTDIR)/ocamlc$(EXE)` _runs_ on `$(OCAMLRUN)` but produces code for `$(ROOTDIR)/runtime/ocamlrun$(EXE)` and so although it's doing `-I $(ROOTDIR)/stdlib`, it still uses `$(OCAMLRUN)`.

dra27 commented on Makefile:
```diff
@@ -1045,7 +1045,7 @@ toplevel/native/topeval.cmx: otherlibs/dynlink/dynlink.cmxa
 
 bytecomp/opcodes.ml: runtime/caml/instruct.h $(make_opcodes)
-	runtime/ocamlrun$(EXE) $(make_opcodes) -opcodes < $< > $@
+	$(NEW_OCAMLRUN) $(make_opcodes) -opcodes < $< > $@
```
cf. https://github.com/ocaml/ocaml/pull/10293. This alteration doesn't affect that problem (either good or bad). Note though that I have as part of fixing something else implemented the suggestion in https://github.com/ocaml/ocaml/pull/10293#issuecomment-797732980 of moving `tools/make_opcodes` into `ocamlc` (so `boot/ocamlc` is capable of doing this) which might simplify things further?

dra27 reviewed (commented):
I hope AppVeyor won't make me eat my words, but this looks fine to me. I went through commit-by-commit. I think it's worth adding comments to clarify why certain runtimes are selected (for `make runtop` and for `tools/caml-tex` especially).

It _might_ be worth holding off on the `NEW_OCAMLRUN` variable change, only because that code _might_ be deleted entirely by a PR of mine, but that's entirely up to you, given that the work I'm doing is both not yet opened in a PR itself or approved!

dra27 commented:
> > Just a quick note the failure on the full-flambda job is unrelated (and is being worked on)
>
> Thanks! Do you need help there? Something that I can help with?

It's OK, thank you - it's a tweak to the GitHub Actions pipeline to stop the failure to build dune creating a red cross on the PR!

shindere commented:
Thanks a lot, David!

Will read your answe in details later.

Just a quick question: is there any hope that the work you started to do
on integrating FlexDLL better will be in a mergeable state soon?

I really liked it, spent time reviewing it, and now it feels it would
really be nice if that could get integrated because it felt like it
would simplify many things.

Could you also elaborate on what you wrote about your work that might
make part of this PR go away completely? Was that the fact that ocamlc
could generate the opcodes?

dra27 commented:
> Just a quick question: is there any hope that the work you started to do on integrating FlexDLL better will be in a mergeable state soon?

Yes - it's very high on my list, sorry!

> Could you also elaborate on what you wrote about your work that might make part of this PR go away completely? Was that the fact that ocamlc could generate the opcodes?

Indeed - instead of `tools/make_opcodes` you have `boot/ocamlc -bopcodes`

shindere commented:
David Allsopp (2021/04/22 05:28 -0700):
> > Just a quick question: is there any hope that the work you started to do on integrating FlexDLL better will be in a mergeable state soon?
>
> Yes - it's very high on my list, sorry!

Sounds good! Don't hesitate to ping me to re-review!

> > Could you also elaborate on what you wrote about your work that might make part of this PR go away completely? Was that the fact that ocamlc could generate the opcodes?
>
> Indeed - instead of `tools/make_opcodes` you have `boot/ocamlc
> -bopcodes`

Would be happy to review that one, too, so don't hesitate to mention me
so that I'm notified when you open it!

shindere commented:
David Allsopp (2021/04/22 04:35 -0700):
> I hope AppVeyor won't make me eat my words, but this looks fine to me.
> I went through commit-by-commit. I think it's worth adding comments to
> clarify why certain runtimes are selected (for `make runtop` and for
> `tools/caml-tex` especially).

Are you suggesting the comments because these runtimes were changed? For
`camltex` at least I'm pretty sure it was not intentionnal to use
`runtime/ocaml`. What do you think, @Octachron?

> It _might_ be worth holding off on the `NEW_OCAMLRUN` variable change,
> only because that code _might_ be deleted entirely by a PR of mine,
> but that's entirely up to you, given that the work I'm doing is both
> not yet opened in a PR itself or approved!

Well then I'd say let's keep it so that we have the requested
flexibility to interact with the build system as long as we need it. We
can always remove this variable when yoour PR lands. Especially given
that it's a private variable, it's not in `Makefiile.config` so IMO it
really doesn't hurt to have it. Not that I like it particularly, I would
have preferred to have `BOOT_OCAMLRUN` pointing to boot/ocamlrun and
`OCAMLRUN` pointing to `runtime/ocamlrun`, but that did not quite work.

> Therefore, `ocaml$(EXE)` is produced using the `BOOT_OCAMLC` and
> should therefore be running on `$(OCAMLRUN)` and not
> `runtime/ocamlrun$(EXE)`.

Thanks for having thought about it!

> I think the conclusion of my ramble is that above the `$(MAKE)
> coldstart)` in the `runtop` recipe it might be worth adding a comment
> that `$(RUNTOP)` could want changing to use `runtime/ocamlrun` if we
> ever _stop_ doing `coldstart` at the start of `runtop`.

I confess I'm a bit skeptical about the principle to add comment saying
"if you change this then think aobut that". But if you insist and sayit's really important to you I'll do. :-)


>
> > @@ -319,7 +319,7 @@ caml_tex := caml-tex$(EXE)
>
>  $(caml_tex): INCLUDES += $(addprefix -I $(ROOTDIR)/otherlibs/,str $(UNIXLIB))
>  $(caml_tex): $(caml_tex_files)
> -	$(ROOTDIR)/runtime/ocamlrun$(EXE) $(ROOTDIR)/ocamlc$(EXE) -nostdlib \
> +	$(OCAMLRUN) $(ROOTDIR)/ocamlc$(EXE) -nostdlib \
>
> Likewise, this is correct but possibly worth a comment clarifying
> (especially if what I write is wrong!). `$(ROOTDIR)/ocamlc$(EXE)`
> _runs_ on `$(OCAMLRUN)` but produces code for
> `$(ROOTDIR)/runtime/ocamlrun$(EXE)` and so although it's doing `-I
> $(ROOTDIR)/stdlib`, it still uses `$(OCAMLRUN)`.

Well I hope ultimately this line will be simplified and there will just
be a call to `$(OCAMLC)`, with `OCAMLC` correctly defined.

Concretely, what owuld you like me to do so that this can get in?

Thanks!

dra27 commented:
I was suggesting the comments because the reasons why those calls are different is potentially confusing - the lines looked clearly different before because they had hard-coded (if wrong!) runtimes:

- `$(RUNTOP)` is using the _boot_ runtime but if you use it it will pick _new_ primitives you just added in `runtime/` which is behaviour you want, but is a possibly accidental consequence of it having `$(MAKE) coldstart`. If you don't like the style of warning comment, then perhaps putting `# The runtop recipe begins with `$(MAKE)` coldstart` and so `boot/ocamlrun` and `runtime/ocamlrun` will be the same`. Of course the 500+ lines between the definition of `RUNTOP` and its single use isn't ideal for spotting the subtle link between the two lines!
- For caml-tex, having looked again at where it's invoked, I think there really is a "mistake" - `caml-tex` itself is run using `boot/ocamlrun`, so I'm not sure there's any reason not to compile it with `-I $(ROOTDIR)/boot` instead of `-I $(ROOTDIR)/stdlib` and therefore just `$(CAMLC)`.

No problem for me for the `NEW_OCAMLRUN` change to go in now.

shindere commented:
David Allsopp (2021/04/22 08:31 -0700):
> I was suggesting the comments because the reasons why those calls are different is potentially confusing - the lines looked clearly different before because they had hard-coded (if wrong!) runtimes:
>
> - `$(RUNTOP)` is using the _boot_ runtime but if you use it it will
> pick _new_ primitives you just added in `runtime/` which is behaviour
> you want, but is a possibly accidental consequence of it having
> `$(MAKE) coldstart`. If you don't like the style of warning comment,
> then perhaps putting `# The runtop recipe begins with `$(MAKE)`
> coldstart` and so `boot/ocamlrun` and `runtime/ocamlrun` will be the
> same`. Of course the 500+ lines between the definition of `RUNTOP` and
> its single use isn't ideal for spotting the subtle link between the
> two lines!

Since RUNTOP is used only once, perhaps we should simply get rid of it?
Do you think it would be prefer to keep running it on the same runtime
as before?

> - For caml-tex, having looked again at where it's invoked, I think
> there really is a "mistake" - `caml-tex` itself is run using
> `boot/ocamlrun`, so I'm not sure there's any reason not to compile
> it with `-I $(ROOTDIR)/boot` instead of `-I $(ROOTDIR)/stdlib` and
> therefore just `$(CAMLC)`.

I think that would make it more similar to how other tools like ocamldoc
or ocamltest are compiled, which makes sense to me. I don't see why
caml-tex would be special in that way.

So my understanding is that the only two remaining points to agree upon
are this RUNTOP thing and caml-tex. Right?

dra27 commented:
> I was suggesting the comments because the reasons why those calls are different is potentially confusing - the lines looked clearly different before because they had hard-coded (if wrong!) runtimes: - `$(RUNTOP)` is using the _boot_ runtime but if you use it it will pick _new_ primitives you just added in `runtime/` which is behaviour you want, but is a possibly accidental consequence of it having `$(MAKE) coldstart`. If you don't like the style of warning comment, then perhaps putting `# The runtop recipe begins with `$(MAKE)` coldstart` and so `boot/ocamlrun` and `runtime/ocamlrun` will be the same`. Of course the 500+ lines between the definition of `RUNTOP` and its single use isn't ideal for spotting the subtle link between the two lines!
>
>> Since RUNTOP is used only once, perhaps we should simply get rid of it? Do you think it would be prefer to keep running it on the same runtime as before?

I'd happily see `$(RUNTOP)` gone, yes! Changing the runtime as you have is definitely correct. My point is that when a developer runs `make runtop` they are expecting to get the the `runtime/ocamlrun` (i.e. with changes they may have worked on). However, `ocaml` in the tree is always compiled to run with `boot/ocamlrun`. The point I think worth clarifying (because it's not obvious) is that the `coldstart` at the start means that `boot/ocamlrun` will always be the same as `runtime/ocamlrun` when you do `make runtop`.

> > - For caml-tex, having looked again at where it's invoked, I think there really is a "mistake" - `caml-tex` itself is run using `boot/ocamlrun`, so I'm not sure there's any reason not to compile it with `-I $(ROOTDIR)/boot` instead of `-I $(ROOTDIR)/stdlib` and therefore just `$(CAMLC)`.

> I think that would make it more similar to how other tools like ocamldoc or ocamltest are compiled, which makes sense to me. I don't see why caml-tex would be special in that way. So my understanding is that the only two remaining points to agree upon are this RUNTOP thing and caml-tex. Right?

Indeed - although I think we're in agreement about `caml-tex` that it should be using the same stdlib as the other tools?

EduardoRFS commented:
Just a happy comment, this patch also solves a problem related to cross compiling, where we need to override the `runtime/ocamlrun`, now it is just another variable that can be changed on make.

shindere commented:
Eduardo Rafael (2021/04/23 15:26 -0700):
> Just a happy comment, this patch also solves a problem related to
> cross compiling, where we need to override the `runtime/ocamlrun`, now
> it is just another variable that can be changed on make.

Thanks a lot Eduardo! That's indeed what inspired this patch and I plan
to do others in the same vein so that the compilers and stdlib to use
can be given as parameters to the build system. If you have other
suggestions like this, please do not hesitate to share because I'd
definitely prioritize things that are also useful to others.

Many thanks for your support and for having taken the time to leave this
comment and give this positive feedback!

shindere committed:
Clarify the way `make runtop` works.

This also introduces a bit of refactoring between the runtop and
natruntop targets, ensuring both toplevels are called with the same
command-line options.

shindere commented:
David Allsopp (2021/04/23 06:04 -0700):
> > I was suggesting the comments because the reasons why those calls are different is potentially confusing - the lines looked clearly different before because they had hard-coded (if wrong!) runtimes: - `$(RUNTOP)` is using the _boot_ runtime but if you use it it will pick _new_ primitives you just added in `runtime/` which is behaviour you want, but is a possibly accidental consequence of it having `$(MAKE) coldstart`. If you don't like the style of warning comment, then perhaps putting `# The runtop recipe begins with `$(MAKE)` coldstart` and so `boot/ocamlrun` and `runtime/ocamlrun` will be the same`. Of course the 500+ lines between the definition of `RUNTOP` and its single use isn't ideal for spotting the subtle link between the two lines!
> >
> >> Since RUNTOP is used only once, perhaps we should simply get rid of it? Do you think it would be prefer to keep running it on the same runtime as before?
>
> I'd happily see `$(RUNTOP)` gone, yes! Changing the runtime as you
> have is definitely correct. My point is that when a developer runs
> `make runtop` they are expecting to get the the `runtime/ocamlrun`
> (i.e. with changes they may have worked on). However, `ocaml` in the
> tree is always compiled to run with `boot/ocamlrun`. The point I think
> worth clarifying (because it's not obvious) is that the `coldstart` at
> the start means that `boot/ocamlrun` will always be the same as
> `runtime/ocamlrun` when you do `make runtop`.

I documented that and also did a bit of refactoring between the `runtop`
and `natruntop` targets. The `TOPFLAGS` variable didn't look that
useful so I simply re-used it ot factorize the flags passed to both
toplevels. I can rename it to e.g. `OC_TOPFLAGS` and add a reference to
`TOPFLAGS` if it is felt that there is a need for users to be able to
pass flags to the toplevel. I am also wondering about the usefulness of
silencing the commands used to invoke the toplevels. Should'nt we get
rid of these `@` signs and let the users see what exactly is called?

> > > - For caml-tex, having looked again at where it's invoked, I think there really is a "mistake" - `caml-tex` itself is run using `boot/ocamlrun`, so I'm not sure there's any reason not to compile it with `-I $(ROOTDIR)/boot` instead of `-I $(ROOTDIR)/stdlib` and therefore just `$(CAMLC)`.
>
> > I think that would make it more similar to how other tools like ocamldoc or ocamltest are compiled, which makes sense to me. I don't see why caml-tex would be special in that way. So my understanding is that the only two remaining points to agree upon are this RUNTOP thing and caml-tex. Right?
>
> Indeed - although I think we're in agreement about `caml-tex` that it
> should be using the same stdlib as the other tools?

Yes.

I just rebased and changed the command used to compile `caml-tex`.

How far are we left from that being merged?

gasche commented:
> I can rename it to e.g. `OC_TOPFLAGS` and add a reference to
`TOPFLAGS` if it is felt that there is a need for users to be able to
pass flags to the toplevel.

I regularly use

```
make runtop -j TOPFLAGS="-dlambda"
```

Is there a way to do this with the new code?

(I guess that I should document this use of TOPFLAGS in HACKING.adoc, which may help other users of the toplevel or other people refactoring the build system.)

shindere commented:
Gabriel Scherer (2021/04/26 07:22 -0700):
> > I can rename it to e.g. `OC_TOPFLAGS` and add a reference to
> `TOPFLAGS` if it is felt that there is a need for users to be able to
> pass flags to the toplevel.
>
> I regularly use
>
> ```
> make runtop -j TOPFLAGS="-dlambda"
> ```
>
> Is there a way to do this with the new code?

I just amended the commit in question. The command you mentionned above should
continue to work exactly as before. Would you mind giving it a try,
please?

shindere commented:
Gabriel Scherer (2021/04/26 07:22 -0700):
> (I guess that I should document this use of TOPFLAGS in HACKING.adoc,
> which may help other users of the toplevel or other people refactoring
> the build system.)

We could add a commentary to the Makefile, at least, at the moment.

In the long run, it could make sense, yes, to gather in one place all
the varialbes that users can use to affect the build system, once this
has all made a bit more coherent (we lack uniformity at the moment,
different sub-makefiles may use different variables; as you may have
noticed I started to work on this.)

dra27 commented:
From my perspective it "looks" good to go, but there's a subtlety about compiling caml-tex which my previous comments had overlooked. It's correct to be _running_ the compiler with `boot/ocamlrun` to build caml-tex, but it's not correct to use `boot/ocamlc` to do so - it's got the wrong `config.cmo` and because `caml-tex` uses `str.cma` and `unix.cma` this matters. The Windows build on AppVeyor fails because `ext_dll` is "wrong" in the boot compiler and the i386-static job fails because `supports_shared_libraries` is "wrong" in the boot compiler. So it can use `$(OCAMLRUN) $(ROOTDIR)/ocamlc` but it can't use `$(CAMLC)`

shindere commented:
David Allsopp (2021/04/26 07:45 -0700):
> From my perspective it "looks" good to go, but there's a subtlety
> about compiling caml-tex which my previous comments had overlooked.
> It's correct to be _running_ the compiler with `boot/ocamlrun` to
> build caml-tex, but it's not correct to use `boot/ocamlc` to do so -
> it's got the wrong `config.cmo` and because `caml-tex` uses `str.cma`
> and `unix.cma` this matters. The Windows build on AppVeyor fails
> because `ext_dll` is "wrong" in the boot compiler and the i386-static
> job fails because `supports_shared_libraries` is "wrong" in the boot
> compiler. So it can use `$(OCAMLRUN) $(ROOTDIR)/ocamlc` but it can't
> use `$(CAMLC)`

So, are you actually saying that the change like it was before was
right? I'm sorry, I'm not following you...

dra27 commented:
> From my perspective it "looks" good to go, but there's a subtlety about compiling caml-tex which my previous comments had overlooked. It's correct to be _running_ the compiler with `boot/ocamlrun` to build caml-tex, but it's not correct to use `boot/ocamlc` to do so - it's got the wrong `config.cmo` and because `caml-tex` uses `str.cma` and `unix.cma` this matters. The Windows build on AppVeyor fails because `ext_dll` is "wrong" in the boot compiler and the i386-static job fails because `supports_shared_libraries` is "wrong" in the boot compiler. So it can use `$(OCAMLRUN) $(ROOTDIR)/ocamlc` but it can't use `$(CAMLC)`
>
> > So, are you actually saying that the change like it was before was right? I'm sorry, I'm not following you...

Yes, we just have even more reasons to comment the command! I think it wants to be:

```make
# caml-tex uses str.cma and unix.cma and so must be compiled with
# $(ROOTDIR)/ocamlc not $(ROOTDIR)/boot/ocamlc since the boot
# compiler does not necessarily have the correct shared library
# configuration.
$(caml_tex): $(caml_tex_files)
	$(OCAMLRUN) $(ROOTDIR)/ocamlc$(EXE) -nostdlib \
	                            -I $(ROOTDIR)/stdlib $(LINKFLAGS) -linkall \
	                            -o $@ -no-alias-deps $^
```

shindere committed:
Build system: simplify the compilation of the caml-tex tool

shindere commented:
Thanks a lot, David.

I just pushed your suggestion, let's see how that goes.

dra27 commented on Makefile:
```diff
@@ -621,25 +615,30 @@ ocaml.tmp: $(ocaml_dependencies)
 partialclean::
 	rm -f ocaml$(EXE)
 
+OC_TOPFLAGS = -nostdlib -I stdlib -I toplevel -noinit $(TOPINCLUDES) $(TOPFLAGS)
```
Very pedantically; it's not as if we use `runtop` with `--warn-undefined-variables`:
```suggestion
TOPFLAGS ?=
OC_TOPFLAGS = -nostdlib -I stdlib -I toplevel -noinit $(TOPINCLUDES) $(TOPFLAGS)
```

dra27 reviewed (approved):
Assuming CI agrees, this is good to go - the `Changes` entry will need an update and I made one pedantic suggestion to have a default for `TOPFLAGS`

shindere commented:
All done. Rebased, Changes entry updated and `TOPFLAGS` initialized as
wisely suggested, even included a comment to document its use.

dra27 commented:
check-typo is grumbling about `tools/Makefile`

shindere commented:
Should be fixed! Thanks!

shindere commented:
David Allsopp (2021/04/27 23:30 -0700):
> Merged #10366 into trunk.

Thank you!!!
